src/Compilers/CSharp/Portable/BoundTree/Statement.cs

```diff
@@ -18,25 +18,36 @@ partial class BoundStatement : IStatement
 
         protected abstract OperationKind StatementKind { get; }
 
-        public abstract void Accept(IOperationVisitor visitor);
+        public abstract void Accept(OperationVisitor visitor);
 
-        public abstract TResult Accept<TArgument, TResult>(IOperationVisitor<TArgument, TResult> visitor, TArgument argument);
+        public abstract TResult Accept<TArgument, TResult>(OperationVisitor<TArgument, TResult> visitor, TArgument argument);
     }
 
     partial class BoundBlock : IBlockStatement
     {
-        ImmutableArray<IStatement> IBlockStatement.Statements => this.Statements.As<IStatement>();
+        private static readonly ConditionalWeakTable<BoundBlock, object> s_blockStatementsMappings =
+            new ConditionalWeakTable<BoundBlock, object>();
+
+        ImmutableArray<IStatement> IBlockStatement.Statements
+        {
+            get
+            {
+                return (ImmutableArray<IStatement>) s_blockStatementsMappings.GetValue(this,
+                    blockStatement => { return blockStatement.Statements.AsImmutable<IStatement>().WhereAsArray(statement => statement.Kind != OperationKind.None); }
```

Comment:
@cston went through an exercise in removing all CWT uses from IOperation implementations. Did we resolve whether or not to adopt that position?